### PR TITLE
New implementation of Dask & Prefect executors

### LINKED ISSUE #116: Improve serializability of Recipe and related classes
After the refactor in #101, our dask graphs have gotten a lot bigger.

I did the following, based on the NOAA OISST recipe from the docs

```python
import pandas as pd
import s3fs
import gcsfs
import json

from pangeo_forge.patterns import pattern_from_file_sequence
from pangeo_forge.recipes import XarrayZarrRecipe
from pangeo_forge.storage import FSSpecTarget, CacheFSSpecTarget

# set up recipe

input_url_pattern = (
    "https://www.ncei.noaa.gov/data/sea-surface-temperature-optimum-interpolation"
    "/v2.1/access/avhrr/{yyyymm}/oisst-avhrr-v02r01.{yyyymmdd}.nc"
)

dates = pd.date_range("1981-09-01", "2021-01-05", freq="D")
input_urls = [
    input_url_pattern.format(
        yyyymm=day.strftime("%Y%m"), yyyymmdd=day.strftime("%Y%m%d")
    )
    for day in dates
]

pattern = pattern_from_file_sequence(input_urls, "time", nitems_per_file=1)
recipe = XarrayZarrRecipe(pattern, inputs_per_chunk=5)

# set up storage

fmt = 'zarr'
recipe_key = 'oisst_avhrr_v02r01'
recipe_name = f'{feedstock_name}/{recipe_key}'

feedstock_name = 'noaa_oisst'

cache_base = f'gs://pangeo-forge-us-central1/pangeo-forge-cache'
metadata_base = f'gs://pangeo-forge-us-central1/pangeo-forge-metadata'

with open('pangeo-181919-cc01a4fbb1ef.json') as fp:
    token = json.load(fp)
fs_gcs = gcsfs.GCSFileSystem(token=token, cache_timeout=-1, cache_type="none")

endpoint_url = 'https://ncsa.osn.xsede.org'
fs_osn = s3fs.S3FileSystem(
    key=osn_key,
    secret=osn_secret,
    client_kwargs={'endpoint_url': endpoint_url},
    default_cache_type='none',
    default_fill_cache=False,
    listings_expiry_time=0.01,
)

target_base = f's3://Pangeo/pangeo-forge'
recipe.input_cache = CacheFSSpecTarget(fs_gcs, f"{cache_base}/{recipe_name}")
recipe.metadata_cache = FSSpecTarget(fs_gcs, f"{metadata_base}/{recipe_name}")
recipe.target = FSSpecTarget(fs_osn, f"{target_base}/{recipe_name}.{fmt}") 
```

Then I converted it to a dask delayed object to execute on a Dask Gateway cluster

```python
from pangeo_forge.executors import PythonPipelineExecutor
executor_dask = DaskPipelineExecutor()
pl = recipe.to_pipelines()
plan = executor_dask.pipelines_to_plan(pl)  # dask.Delayed obj
```

The graph is about 10 MB, not huge, right?
```python
import cloudpickle
serialized = cloudpickle.dumps(plan.dask)
len(serialized)
# -> 10620079
```

But when I call
```python
dask.persist(plan, retries=5)
```

I quickly run out of memory in my notebook pod (16 GB) before the the `persist` call finished. I also tried with `, optimize_graph=False`.

One culprit could be the new FilePattern class, which stores a big xarray dataset internally
https://github.com/pangeo-forge/pangeo-forge/blob/1c5f3520d6fdca718cb372d60161d4fd8617dcbb/pangeo_forge/patterns.py#L81

I feel like we could optimize this using `__getstate__` / `__setstate`.

@martindurant it would be great if you could look at this.

### LINKED ISSUE #151: Dask distributed client memory overruns when with large time dimensions.
I performed some successful initial tests processing the [noaa-oisst](https://github.com/pangeo-forge/staged-recipes/pull/36) recipe using a smaller time dimension. When extending the recipe to use the full time range `pd.date_range("1981-09-01", "2021-01-05", freq="D")` I am seeing memory overruns on the container where the Prefect Flow is being executed (the container is killed for exceeding its 30GB hard limit).  

To obtain some more details I attempted to run the Flow in a local container directly using `flow.run` rather than the Prefect Agent's subprocess.  I added some `tracemalloc` instrumentation and found rapidly growing memory utilization from `msgpack`.
```
INFO:root:/srv/conda/envs/notebook/lib/python3.8/site-packages/msgpack/__init__.py:0: size=66.9 MiB, count=6, average=11.2 MiB
INFO:root:/srv/conda/envs/notebook/lib/python3.8/site-packages/msgpack/__init__.py:0: size=1156 MiB, count=120, average=9868 KiB
INFO:root:/srv/conda/envs/notebook/lib/python3.8/site-packages/msgpack/__init__.py:0: size=2408 MiB, count=251, average=9826 KiB
INFO:root:/srv/conda/envs/notebook/lib/python3.8/site-packages/msgpack/__init__.py:0: size=3603 MiB, count=377, average=9787 KiB
```
Before the process was killed after exceeding the container's 5GB hard limit.

A traceback of the largest block prior to the container being killed points to message serialization in distributed
```
INFO:root:34 memory blocks: 1448426.9 KiB
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/threading.py", line 890
INFO:root:    self._bootstrap_inner()
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/threading.py", line 932
INFO:root:    self.run()
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/threading.py", line 870
INFO:root:    self._target(*self._args, **self._kwargs)
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/concurrent/futures/thread.py", line 80
INFO:root:    work_item.run()
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/concurrent/futures/thread.py", line 57
INFO:root:    result = self.fn(*self.args, **self.kwargs)
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/utils.py", line 1471
INFO:root:    return await loop.run_in_executor(_offload_executor, lambda: fn(*args, **kwargs))
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/comm/utils.py", line 31
INFO:root:    protocol.dumps(
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/distributed/protocol/core.py", line 70
INFO:root:    frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
INFO:root:  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/msgpack/__init__.py", line 35
INFO:root:    return Packer(**kwargs).pack(o)
Killed
```
As a note, my scheduler and workers continue operating fine and processing tasks within their smaller memory limits until the Prefect flow runner is killed.

I've engaged with Prefect support on Slack but this appears to be a potential distributed client issue and I don't know about the depth of their dask expertise now that Jim will be leaving.  I'm very inexperienced with dask internals so any suggestions or advice here would be greatly appreciated 🙇  cc @rabernat @TomAugspurger .

### PR DESCRIPTION
**Pull request by rabernat**
possibly fixes #116
maybe eventually fixes #151

### CODE REVIEW

File: pangeo_forge_recipes/recipes/base.py [hunk 1]
```diff
@@ -95,6 +93,26 @@ def __post_init__(self):
...
+    def to_dask(self):
+        cls_ = self.__class__
+
+        def self_task():
+            return self
+
+        self_ = delayed(self_task)()
+
+        layers = [
+            [delayed(cls_.cache_input)(self_, input_key) for input_key in self.iter_inputs()],
```
Comment (rabernat) [inline pangeo_forge_recipes/recipes/base.py:105]: Dask experts @TomAugspurger / @martindurant: does this look right / reasonable? Is there a way to make this a HighLevelGraph layer?

I tried using dask bag, e.g. 
```python
db.from_sequence(range(self.iter_inputs()).map(cls_.cache_input, self_)
```

but I couldn't get around the fact that the _non-mapped_ argument (`self_` has to come after the mapped one in the function signature.

Comment (TomAugspurger) [reply pangeo_forge_recipes/recipes/base.py:105]: This feels overly complicated. Give me a bit to play with it.

**Review by TomAugspurger — APPROVED** [review]
Feel free to merge when you're ready. I haven't been able to make a less hacky version. If I can I'll make a PR later.

**Comment by rabernat** [comment]
Question: can I just remove the `to_piplines()` method? Or do we need a deprecation cycle?

**Comment by TomAugspurger** [comment]
From the azure bakery side we're fine to just remove it. Do we know of everyone using this already?

**Pull request closed without merging**